src/chainstate/stacks/index/proofs.rs

```diff
@@ -0,0 +1,1229 @@
+            TrieMerkleProofType::Node48((ref chr, ref node, ref hashes)) => write!(f, "TrieMerkleProofType::Node48(0x{:02x}, node={:?}, hashes={})", chr, node, hashes_fmt(hashes)),
+            TrieMerkleProofType::Node256((ref chr, ref node, ref hashes)) => write!(f, "TrieMerkleProofType::Node256(0x{:02x}, node={:?}, hashes={})", chr, node, hashes_fmt(hashes)),
+            TrieMerkleProofType::Leaf((ref chr, ref node)) => write!(f, "TrieMerkleProofType::Leaf(0x{:02x}, node={:?})", chr, node),
+            TrieMerkleProofType::Shunt((ref idx, ref hashes)) => write!(f, "TrieMerkleProofType::Shunt(idx={}, hashes={:?})", idx, hashes)
+        }
+    }
+}
+
+impl PartialEq for TrieMerkleProofType {
+    fn eq(&self, other: &TrieMerkleProofType) -> bool {
+        match (self, other) {
+            (TrieMerkleProofType::Node4((ref chr, ref node, ref hashes)), TrieMerkleProofType::Node4((ref other_chr, ref other_node, ref other_hashes))) => {
+                chr == other_chr && node == other_node && slice_partialeq(hashes, other_hashes)
+            },
+            (TrieMerkleProofType::Node16((ref chr, ref node, ref hashes)), TrieMerkleProofType::Node16((ref other_chr, ref other_node, ref other_hashes))) => {
+                chr == other_chr && node == other_node && slice_partialeq(hashes, other_hashes)
+            },
+            (TrieMerkleProofType::Node48((ref chr, ref node, ref hashes)), TrieMerkleProofType::Node48((ref other_chr, ref other_node, ref other_hashes))) => {
+                chr == other_chr && node == other_node && slice_partialeq(hashes, other_hashes)
+            },
+            (TrieMerkleProofType::Node256((ref chr, ref node, ref hashes)), TrieMerkleProofType::Node256((ref other_chr, ref other_node, ref other_hashes))) => {
+                chr == other_chr && node == other_node && slice_partialeq(hashes, other_hashes)
+            },
+            (TrieMerkleProofType::Leaf((ref chr, ref node)), TrieMerkleProofType::Leaf((ref other_chr, ref other_node))) => {
+                chr == other_chr && node == other_node
+            },
+            (TrieMerkleProofType::Shunt((ref idx_1, ref hashes_1)), TrieMerkleProofType::Shunt((ref idx_2, ref hashes_2))) => {
+                idx_1 == idx_2 && hashes_1 == hashes_2
+            },
+            (_, _) => false
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct TrieMerkleProof(Vec<TrieMerkleProofType>);
+
+impl Deref for TrieMerkleProof {
+    type Target = Vec<TrieMerkleProofType>;
+    fn deref(&self) -> &Vec<TrieMerkleProofType> {
+        &self.0
+    }
+}
+
+impl TrieMerkleProof {
+
+    fn make_proof_hashes(node: &TrieNodeType, all_hashes: &Vec<TrieHash>, chr: u8) -> Result<Vec<TrieHash>, Error> {
+        let mut hashes = vec![];
+        assert!(all_hashes.len() == node.ptrs().len());
+
+        for i in 0..node.ptrs().len() {
+            if node.ptrs()[i].id() == TrieNodeID::Empty {
+                hashes.push(TrieHash::from_data(&[]));
```

Comment:
Why do we keep all this empty nodes in our proofs?

Reply:
Should we at least cache `TrieHash::from_data(&[])`?

Reply:
The prover needs to supply sibling hashes, even if they are the "empty" hash.  However, when the prover goes to _encode_ the proof to transmit over the network, it can (and should) compress the the proof to eliminate the redundant hashes.  But if it's okay with you, I'd like to address this in a separate PR :)